server/aesdsocket.c

```diff
@@ -0,0 +1,332 @@
+			daemonFlag = 1;
+	}
+
+
+    // ********************* make socket
+    fd[FD_SOCKET] = socket(PF_INET, SOCK_STREAM, 0);
+
+    // ********************* check socket creation
+    if(fd[FD_SOCKET] < 0)
+    {
+        printf("ERROR: Failed to create socket, exiting...");
+        exit (EXIT_FAILURE);
+    }
+
+    socketInfo.ai_family = PF_INET;
+	socketInfo.ai_socktype = SOCK_STREAM;
+	socketInfo.ai_flags = AI_PASSIVE;
+
+    // ********************* call getaddrinfo and check return_value
+    int return_value = getaddrinfo(NULL, PORT, &socketInfo, &sockAddrInfo);
+    if(return_value != 0)
+	{
+		printf("ERROR: failed to get address information, exiting... %s\n", gai_strerror(return_value));
+		freeaddrinfo(sockAddrInfo);
+		exit (EXIT_FAILURE);
+	}
+
+    // ********************* bind
+    return_value = bind(fd[FD_SOCKET], sockAddrInfo->ai_addr, sizeof(struct sockaddr));
+	if(return_value == -1)
+	{
+		printf("ERROR: failed to bind, exiting... %s\n", strerror(errno));
+		freeaddrinfo(sockAddrInfo);
+		exit (EXIT_FAILURE);
+	}
+
+    freeaddrinfo(sockAddrInfo);
+
+	return_value = listen(fd[FD_SOCKET], MAX_CONNECTION);
+	if(return_value == -1)
+	{
+		printf("ERROR: failed to listen, exiting...\n");
+		exit (EXIT_FAILURE);
+	}
+
+    // ********************* open write file
+	fd[FD_DATA] = open(FILE_OUT_PATH, O_CREAT | O_APPEND | O_RDWR, S_IRWXU);
+    if(fd[FD_DATA] == -1)
+	{
+		printf("ERROR: failed to open file, %s \n", FILE_OUT_PATH);
+		exit (EXIT_FAILURE);
+	}
+
+	sigemptyset(&sigSetOne);
+	sigaddset(&sigSetOne, SIGINT);
+	sigaddset(&sigSetOne, SIGTERM);
+
+    buffer_rx = (char *)malloc(BUFFER_SIZE*sizeof(char));
+    buffer_tx = (char *)malloc(BUFFER_SIZE*sizeof(char));
+	if(buffer_rx == NULL || buffer_tx == NULL)
+	{
+		printf("ERROR: failed to malloc buffers, exiting...\n");
+		exit (EXIT_FAILURE);
+	}	
+
+    if(daemonFlag)
+	{
+		pid = fork();
+
+		if(pid == -1)
+			exit (EXIT_FAILURE);
+
+		else if(pid != 0)
+			exit(0);
+
+		if(setsid() == -1)
+			exit (EXIT_FAILURE);
+
+		if(chdir("/") == -1)
+			exit (EXIT_FAILURE);
+
+		open("/dev/null", O_RDWR);
+
+		dup(0);
+		dup(0);
+	}
+	
+	while(1)
+	{
+		/* Get the accepted client fd */
+		fd[FD_CLIENT] = accept(fd[FD_SOCKET], (struct sockaddr *)&clientSockAddress, &addressSize); 
+
+		/* Error occurred in accept, exit (EXIT_FAILURE) on error */
+		if(fd[FD_CLIENT] == -1)
+		{
+			printf("ERROR: failed to accept client fd, exiting...\n");
+			free(buffer_rx);
+			exit (EXIT_FAILURE);
+		}
+
+		char *IP = inet_ntoa(clientSockAddress.sin_addr);
+		syslog(LOG_DEBUG, "Accepted connection from %s\n", IP);
+
+		/* Don't accept a signal while recieving data */
+		if((return_value = sigprocmask(SIG_BLOCK, &sigSetOne, &sigSetTwo)) == -1)
+			printf("ERROR: failed to sigprocmask, exiting...\n");
+		
+		/* Initially, write to starting of buffer */
+		bufferIndex = 0;	
+		
+		/* Receive data */
+		while((recievedByteCount = recv(fd[FD_CLIENT], buffer_rx + bufferIndex, BUFFER_SIZE, 0)) > 0)
```

Comment:
@cchoi22915 I think it will be possible to exceed array bounds here.  If the previous iteration had `bufferIndex = bufferCount*BUFFER_SIZE -1`  I think you will attempt to read up to `BUFFER_SIZE` bytes into `buffer_rx+bufferIndex` when only 1 byte is remaining.